### PR TITLE
feat: add bid tables 

### LINKED ISSUE #21: Create bid_requests and bid_status_histories tables
# Create bid_requests and bid_status_histories tables

Create tables to handle apartment bid requests and their status history, tracking the complete lifecycle of a tenant's bid on an apartment.

## Problem
We need to:
- Track bid requests from tenants
- Maintain bid request status history
- Support the complete bid lifecycle
- Enable bid status transitions

### Migration Files

Create new migration:
```bash
hasura migrate create create_bid_tables
```

#### `up.sql`:
```sql
-- Create bid_requests table
CREATE TABLE bid_requests (
    id UUID PRIMARY KEY DEFAULT uuid_generate_v4(),
    apartment_id UUID REFERENCES apartments(id) ON DELETE CASCADE,
    tenant_id TEXT REFERENCES users(id) ON DELETE CASCADE,
    current_status TEXT NOT NULL DEFAULT 'PENDING',
    proposed_price DECIMAL(10,2) NOT NULL CHECK (proposed_price > 0),
    desired_move_in TIMESTAMP WITH TIME ZONE NOT NULL,
    created_at TIMESTAMP WITH TIME ZONE DEFAULT NOW(),
    updated_at TIMESTAMP WITH TIME ZONE DEFAULT NOW(),
    deleted_at TIMESTAMP WITH TIME ZONE,
    CONSTRAINT valid_status CHECK (
        current_status IN (
            'PENDING',
            'VIEWED',
            'APPROVED',
            'CONFIRMED',
            'ESCROW_FUNDED',
            'ESCROW_COMPLETED',
            'CANCELLED'
        )
    )
);

-- Create bid_status_histories table
CREATE TABLE bid_status_histories (
    id UUID PRIMARY KEY DEFAULT uuid_generate_v4(),
    bid_request_id UUID REFERENCES bid_requests(id) ON DELETE CASCADE,
    status TEXT NOT NULL,
    notes TEXT,
    changed_by TEXT REFERENCES users(id) ON DELETE SET NULL,
    created_at TIMESTAMP WITH TIME ZONE DEFAULT NOW()
);

-- Create function to record status changes
CREATE OR REPLACE FUNCTION record_bid_status_change()
RETURNS TRIGGER AS $$
BEGIN
    IF (TG_OP = 'INSERT') OR (OLD.current_status IS DISTINCT FROM NEW.current_status) THEN
        INSERT INTO bid_status_histories (
            bid_request_id,
            status,
            changed_by
        ) VALUES (
            NEW.id,
            NEW.current_status,
            current_setting('hasura.user', true)::text
        );
    END IF;
    RETURN NEW;
END;
$$ LANGUAGE plpgsql;

-- Create trigger for status history
CREATE TRIGGER record_bid_status
    AFTER INSERT OR UPDATE ON bid_requests
    FOR EACH ROW
    EXECUTE FUNCTION record_bid_status_change();

-- Create indexes
CREATE INDEX idx_bid_requests_apartment 
    ON bid_requests(apartment_id);
CREATE INDEX idx_bid_requests_tenant 
    ON bid_requests(tenant_id);
CREATE INDEX idx_bid_requests_status 
    ON bid_requests(current_status);
CREATE INDEX idx_bid_requests_dates 
    ON bid_requests(created_at, desired_move_in);
CREATE INDEX idx_bid_histories_request 
    ON bid_status_histories(bid_request_id);
CREATE INDEX idx_bid_histories_dates 
    ON bid_status_histories(created_at);

-- Add trigger for updated_at
CREATE TRIGGER update_bid_requests_updated_at
    BEFORE UPDATE ON bid_requests
    FOR EACH ROW
    EXECUTE FUNCTION update_updated_at_column();

-- Create function to check for multiple active bids
CREATE OR REPLACE FUNCTION check_active_bids()
RETURNS TRIGGER AS $$
BEGIN
    IF NEW.current_status IN ('PENDING', 'VIEWED', 'APPROVED') THEN
        IF EXISTS (
            SELECT 1 FROM bid_requests
            WHERE tenant_id = NEW.tenant_id
            AND id != NEW.id
            AND current_status IN ('PENDING', 'VIEWED', 'APPROVED')
            AND deleted_at IS NULL
        ) THEN
            RAISE EXCEPTION 'Tenant already has an active bid';
        END IF;
    END IF;
    RETURN NEW;
END;
$$ LANGUAGE plpgsql;

-- Create trigger for active bids check
CREATE TRIGGER check_tenant_active_bids
    BEFORE INSERT OR UPDATE ON bid_requests
    FOR EACH ROW
    EXECUTE FUNCTION check_active_bids();
```

#### `down.sql`:
```sql
-- Drop triggers
DROP TRIGGER IF EXISTS check_tenant_active_bids ON bid_requests;
DROP TRIGGER IF EXISTS record_bid_status ON bid_requests;
DROP TRIGGER IF EXISTS update_bid_requests_updated_at ON bid_requests;

-- Drop functions
DROP FUNCTION IF EXISTS check_active_bids();
DROP FUNCTION IF EXISTS record_bid_status_change();

-- Drop indexes
DROP INDEX IF EXISTS idx_bid_requests_apartment;
DROP INDEX IF EXISTS idx_bid_requests_tenant;
DROP INDEX IF EXISTS idx_bid_requests_status;
DROP INDEX IF EXISTS idx_bid_requests_dates;
DROP INDEX IF EXISTS idx_bid_histories_request;
DROP INDEX IF EXISTS idx_bid_histories_dates;

-- Drop tables
DROP TABLE IF EXISTS bid_status_histories;
DROP TABLE IF EXISTS bid_requests;
```

## Status Flow Diagram
```mermaid
stateDiagram-v2
    [*] --> PENDING: Create Bid
    PENDING --> VIEWED: Owner Views
    VIEWED --> APPROVED: Owner Approves
    APPROVED --> CONFIRMED: Owner Confirms
    CONFIRMED --> ESCROW_FUNDED: Tenant Funds
    ESCROW_FUNDED --> ESCROW_COMPLETED: Owner Completes
    
    PENDING --> CANCELLED: Cancel
    VIEWED --> CANCELLED: Cancel
    APPROVED --> CANCELLED: Cancel
```

### Tenant Permissions
```yaml
bid_requests:
  select:
    filter:
      tenant_id: X-Hasura-User-Id
  insert:
    check:
      tenant_id: X-Hasura-User-Id
  update:
    filter:
      tenant_id: X-Hasura-User-Id
    check:
      current_status:
        _in: ["CANCELLED"]

bid_status_histories:
  select:
    filter:
      bid_request:
        tenant_id: X-Hasura-User-Id
```

### Landlord Permissions
```yaml
bid_requests:
  select:
    filter:
      apartment:
        owner_id: X-Hasura-User-Id
  update:
    filter:
      apartment:
        owner_id: X-Hasura-User-Id
    check:
      current_status:
        _in: ["VIEWED", "APPROVED", "CONFIRMED"]

bid_status_histories:
  select:
    filter:
      bid_request:
        apartment:
          owner_id: X-Hasura-User-Id
```

## Verification Steps
1. Apply migration:
```bash
hasura migrate apply
```

2. Test status flow:
```sql
-- Create bid
INSERT INTO bid_requests (
    apartment_id,
    tenant_id,
    proposed_price,
    desired_move_in
) VALUES (
    'apartment-uuid',
    'tenant-id',
    1000.00,
    NOW() + interval '1 month'
);

-- Update status
UPDATE bid_requests
SET current_status = 'VIEWED'
WHERE id = 'bid-uuid';

-- Verify history
SELECT * FROM bid_status_histories
WHERE bid_request_id = 'bid-uuid'
ORDER BY created_at;
```

## Prerequisites
- [x] Users table created
- [x] Apartments table created
- [ ] Permissions system configured

## Success Criteria
- [ ] Tables created with correct constraints
- [ ] Status history automatically recorded
- [ ] Multiple active bids prevented
- [ ] Indexes optimized for common queries
- [ ] Permissions properly configured
- [ ] Status flow working correctly

## Additional Context
This implementation:
- Enforces status flow rules
- Maintains complete status history
- Prevents multiple active bids
- Records who made each status change
- Supports soft deletion
- Includes proper indexing

## Related Issues
- safetrustcr/Backend#22 Firebase Auth Integration
- safetrustcr/Backend#10 Create Apartments Table

### PR DESCRIPTION
**Pull request by juandiegocv27**
## ❗ **PR Information**
This pull request introduces tables for managing apartment bid requests and tracking bid status history, covering the entire lifecycle of tenant bids.

## 🌀 Summary of Changes
- **Database Tables**:
  - `bid_requests`: Contains fields for `apartment_id`, `tenant_id`, `current_status`, `proposed_price`, `desired_move_in`, and timestamps.
  - `bid_status_histories`: Logs changes with `bid_request_id`, `status`, `notes`, `changed_by`, and `created_at`.
- **Functions & Triggers**:
  - Automatically updates the `updated_at` column.
  - Logs status changes in the history table.
  - Ensures that tenants cannot have multiple active bids.
- **Optimizations**:
  - Added indexes to improve query performance.
  - Implemented constraints and foreign keys to maintain data integrity.

## 📂 Related Issue
- This pull request will **close #21** upon merging.

### CODE REVIEW

File: migrations/safetrust/1732865994413_create_bid_tables/up.sql [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+CREATE OR REPLACE FUNCTION update_updated_at_column()
+RETURNS TRIGGER AS $$
+BEGIN
+    NEW.updated_at = NOW();
+    RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+-- Tabla principal de solicitudes de ofertas
+CREATE TABLE bid_requests (
```
Comment (rvalenciano) [inline migrations/safetrust/1732865994413_create_bid_tables/up.sql:11]: Lets add these tables all in:

/metadata/databases/safetrust/tables/tables.yaml

files so they can be tracked in the graphql endpoints.

https://hasura.io/docs/2.0/migrations-metadata-seeds/manage-migrations/#create-manual-migrations

> The corresponding Hasura Metadata changes, if any, can be made directly in the appropriate Metadata file in the /metadata directory, refer to [Metadata format](https://hasura.io/docs/2.0/migrations-metadata-seeds/manage-migrations/).

Check other tables (users) as a reference

**Review by rvalenciano — CHANGES_REQUESTED** [review]
Lets modify the metadata and we are done

File: migrations/safetrust/1732865994413_create_bid_tables/up.sql [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+CREATE TABLE bid_status_histories (
+    id UUID PRIMARY KEY DEFAULT uuid_generate_v4(),
+    bid_request_id UUID REFERENCES bid_requests(id) ON DELETE CASCADE,
+    status TEXT NOT NULL,
+    notes TEXT,
+    changed_by TEXT REFERENCES users(id) ON DELETE SET NULL,
+    created_at TIMESTAMP WITH TIME ZONE DEFAULT NOW()
+);
+
+-- Función para registrar cambios de estado
```
Comment (rvalenciano) [inline migrations/safetrust/1732865994413_create_bid_tables/up.sql:44]: This needs to either be removed or translated to English

File: migrations/safetrust/1732865994413_create_bid_tables/up.sql [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            NEW.id,
+            NEW.current_status,
+            current_setting('hasura.user', true)::text
+        );
+    END IF;
+    RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+-- Trigger para registrar cambios de estado
```
Comment (rvalenciano) [inline migrations/safetrust/1732865994413_create_bid_tables/up.sql:63]: Same as before

File: migrations/safetrust/1732865994413_create_bid_tables/up.sql [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+CREATE TRIGGER record_bid_status
+    AFTER INSERT OR UPDATE ON bid_requests
+    FOR EACH ROW
+    EXECUTE FUNCTION record_bid_status_change();
+
+-- Índices para optimización
```
Comment (rvalenciano) [inline migrations/safetrust/1732865994413_create_bid_tables/up.sql:69]: Same as before

File: migrations/safetrust/1732865994413_create_bid_tables/up.sql [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+            AND deleted_at IS NULL
+        ) THEN
+            RAISE EXCEPTION 'Tenant already has an active bid';
+        END IF;
+    END IF;
+    RETURN NEW;
+END;
+$$ LANGUAGE plpgsql;
+
+-- Trigger para verificar ofertas activas
```
Comment (rvalenciano) [inline migrations/safetrust/1732865994413_create_bid_tables/up.sql:108]: Same as before

File: migrations/safetrust/1732865994413_create_bid_tables/up.sql [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+CREATE INDEX idx_bid_histories_dates 
+    ON bid_status_histories(created_at);
+
+-- Trigger para actualizar updated_at
+CREATE TRIGGER update_bid_requests_updated_at
+    BEFORE UPDATE ON bid_requests
+    FOR EACH ROW
+    EXECUTE FUNCTION update_updated_at_column();
+
+-- Función para verificar ofertas activas
```
Comment (rvalenciano) [inline migrations/safetrust/1732865994413_create_bid_tables/up.sql:89]: Same as before

**Review by juandiegocv27 — COMMENTED** [review]
already made the changes requested

**Review by rvalenciano — APPROVED** [review]

**Pull request merged**